sway-core/src/asm_generation/fuel/fuel_asm_builder.rs

```diff
@@ -1146,32 +1151,54 @@ impl<'ir, 'eng> FuelAsmBuilder<'ir, 'eng> {
 
     fn compile_load(&mut self, instr_val: &Value, src_val: &Value) -> Result<(), CompileError> {
         let owning_span = self.md_mgr.val_to_span(self.context, *instr_val);
-        if src_val
+        let src_ty = src_val
             .get_type(self.context)
             .and_then(|src_ty| src_ty.get_pointee_type(self.context))
-            .map_or(true, |inner_ty| !self.is_copy_type(&inner_ty))
-        {
-            Err(CompileError::Internal(
-                "Attempt to load from non-copy type.",
-                owning_span.unwrap_or_else(Span::dummy),
-            ))
-        } else {
-            let src_reg = self.value_to_register(src_val)?;
-            let instr_reg = self.reg_seqr.next();
+            .filter(|inner_ty| self.is_copy_type(inner_ty))
+            .ok_or_else(|| {
+                CompileError::Internal(
+                    "Attempt to load from non-copy type.",
+                    owning_span.clone().unwrap_or_else(Span::dummy),
+                )
+            })?;
+        let byte_len = ir_type_size_in_bytes(self.context, &src_ty);
 
-            self.cur_bytecode.push(Op {
-                opcode: Either::Left(VirtualOp::LW(
-                    instr_reg.clone(),
-                    src_reg,
-                    VirtualImmediate12 { value: 0 },
-                )),
-                comment: "load value".into(),
-                owning_span,
-            });
+        let src_reg = self.value_to_register(src_val)?;
+        let instr_reg = self.reg_seqr.next();
 
-            self.reg_map.insert(*instr_val, instr_reg);
-            Ok(())
+        match byte_len {
```

Comment:
The `byte_len` can be only 1 or 8, or? The `match` is then a bit confusing, giving compile error only on `0 | 2..=7` and giving and impression that >8 is also fine `8..`. How about having only 1, 8, and _ giving compile error?

Reply:
Bigger than 8 is also fine. But I simplified a little bit nonetheless..